test_mavros/include/tests/offboard_control.h

```diff
@@ -0,0 +1,492 @@
+					return;
+				}
+				if (theta == 360) {
+					ROS_INFO("Test complete!");
+					ros::shutdown();
+				}
+				loop_rate.sleep();
+				ros::spinOnce();
+			}
+		}
+	}
+
+	/**
+	 * @brief Eight path motion routine
+	 */
+	void eight_path_motion(ros::Rate loop_rate, control_mode mode){
+		geometry_msgs::PoseStamped setpoint;
+		ROS_INFO("Testing...");
+
+		while (ros::ok()) {
+			// starting point
+			if (mode == POSITION) {
+				X = 0.0f;
+				Y = 0.0f;
+				Z = 1.0f;
+				local_pos_sp_pub.publish(ps);
+			}
+			else if (mode == VELOCITY) {
+				VX = 0.0f - current_x;
+				VY = 0.0f - current_y;
+				VZ = 1.0f - current_z;
+				vel_sp_pub.publish(vs);
+			}
+			else if (mode == ACCELERATION) {
+				// TODO
+				return;
+			}
+
+			wait_and_move(ps);
+
+			// motion routine
+			for (int theta = -180; theta <= 180; theta++) {
+				if (mode == POSITION)
+					local_pos_sp_pub.publish(eight_shape(theta));
+				else if (mode == VELOCITY) {
+					setpoint = eight_shape(theta);
+					VX = setpoint.pose.position.x - current_x;
+					VY = setpoint.pose.position.y - current_y;
+					VZ = setpoint.pose.position.z - current_z;
+					vel_sp_pub.publish(vs);
+				}
+				else if (mode == ACCELERATION) {
+					// TODO
+					return;
+				}
+				if (theta == 180) {
+					ROS_INFO("Test complete!");
+					ros::shutdown();
+				}
+				loop_rate.sleep();
+				ros::spinOnce();
+			}
+		}
+	}
+
+	/**
+	 * @brief Ellipse path motion routine
+	 */
+	void ellipse_path_motion(ros::Rate loop_rate, control_mode mode){
+		geometry_msgs::PoseStamped setpoint;
+		ROS_INFO("Testing...");
+
+		while (ros::ok()) {
+			// starting point
+			if (mode == POSITION) {
+				X = 0.0f;
+				Y = 0.0f;
+				Z = 2.5f;
+				local_pos_sp_pub.publish(ps);
+			}
+			else if (mode == VELOCITY) {
+				// This one gets some strange behavior, maybe due to overshoot on velocity controller
+				// TODO: find a way to limit the velocity between points (probably using ros::Rate)
+				VX = 0.0f - current_x;
+				VY = 0.0f - current_y;
+				VZ = 2.5f - current_z;
+				vel_sp_pub.publish(vs);
+			}
+			else if (mode == ACCELERATION) {
+				// TODO
+				return;
+			}
+
+			wait_and_move(ps);
+
+			// motion routine
+			for (int theta = 0; theta <= 360; theta++) {
+				if (mode == POSITION)
+					local_pos_sp_pub.publish(ellipse_shape(theta));
+				else if (mode == VELOCITY) {
+					setpoint = ellipse_shape(theta);
+					VX = setpoint.pose.position.x - current_x;
+					VY = setpoint.pose.position.y - current_y;
+					VZ = setpoint.pose.position.z - current_z;
+					vel_sp_pub.publish(vs);
+				}
+				else if (mode == ACCELERATION) {
+					// TODO
+					return;
+				}
+				if (theta == 360) {
+					ROS_INFO("Test complete!");
+					ros::shutdown();
+				}
+				loop_rate.sleep();
+				ros::spinOnce();
+			}
+		}
+	}
+
+	/**
+	 * @brief Defines the accepted threshold to the destination/target position
+	 * before moving to the next setpoint.
+	 */
+	void wait_and_move(geometry_msgs::PoseStamped target){
```

Comment:
Well it waits (https://github.com/mavlink/mavros/pull/337/files#diff-3e2be3af4f08d8676c481f5b7228ca3dR464) then it moves (https://github.com/mavlink/mavros/pull/337/files#diff-3e2be3af4f08d8676c481f5b7228ca3dR467)